burnman/layer.py

```diff
@@ -0,0 +1,899 @@
+from burnman import geotherm
+import warnings
+
+
+from .material import Material, material_property
+from .mineral import Mineral
+from .composite import Composite
+from .seismic import Seismic1DModel
+
+import matplotlib.pyplot as plt
+
+
+class Layer(object):
+    """
+    The base class for a planetary layer. The user needs to set the following befor properties can be computed
+    1. set_material(), set the material of the layer, e.g. a mineral, solid_solution, or composite
+    2. set_temperature_mode(), either predefine, or set to an adiabat
+    3. set_pressure_mode(), to set the self-consistent pressure (with user-defined option the pressures can be overwritten).
+    To set the self-consistent pressure the pressure at the top and the gravity at the bottom of the layer need to be set.
+    4. make(), computes the self-consistent part of the layer and starts the settings to compute properties within the layer
+    Note that not the entire planet this layer sits in is self-consistent, as the pressure at the top of the layer is a function of the density
+    within the layer (through the gravity). Entire planets can be computed self-consistently with the planet class.
+    Properties will be returned at the pre-defined radius array, although the evaluate() function
+    can take a new defined depthlist and values are interpolated between these (sufficient sampling of the layer
+    is needed for this to be accurate)
+    """
+
+    def __init__(self, name=None, radii=None, verbose=False):
+        self.name = name
+        self.radii = np.sort(radii)
+        self.outer_radius = max(self.radii)
+        self.inner_radius = min(self.radii)
+        self.thickness = self.outer_radius - self.inner_radius
+        self.n_slices = len(self.radii)
+        self.verbose = verbose
+        self._cached = {}
+        self._pressures = None
+        self._temperatures = None
+        self.sublayers = None
+        self.material = None
+        self.pressure_mode = None
+        self.temperature_mode = None
+ 
+ 
+    def __str__(self):
+        """
+        Prints details of the Layer
+        """
+        writing  =  ' Layer ' + self.name + ' made out of ' + str(self.material.name) + ' with ' \
+                        + str(self.temperature_mode) + ' temperatures  and ' + str(self.pressure_mode) + ' pressures \n'
+        return writing
+            
+            
+    def reset(self):
+        """
+        Resets all cached material properties.
+        It is typically not required for the user to call this function.
+        """
+        self._cached = {}
+        self._pressures = None
+        self._temperatures = None
+        self.sublayers = None
+
+    def set_material(self, material):
+        """
+        Set the material of a Layer with a Material
+        """
+        assert(isinstance(material, Material))
+        self.material = material
+        self.reset()
+
+    def set_temperature_mode(self, temperature_mode='adiabat',
```

Comment:
There needs to be some check that temperature_top has been provided for a single layer adiabatic planet. Also a test would be good.